### PR TITLE
Implement I2C using generics instead of trait objects

### LINKED ISSUE #6: Implement I2C using generics instead of trait objects
The virtual I2C and I2C Master are implemented using `&dyn I2C` which does add some overhead and prevents some compiler optimizations. A possible optimization is the implementation using generics `where I: I2C`.

A good start is the virtual SPI implementation.

https://github.com/UPB-CS-OpenSourceUpstream/tock/blob/de221b2a1748b61f03d59ebf6c053514e2a8dba3/capsules/src/virtual_i2c.rs#L44

### PR DESCRIPTION
**Pull request by JADarius**
### Pull Request Overview

This pull request fixes #6. It is still a work in progress, but I need help with the implementation.

### TODO or Help Wanted

This pull request still needs code review for what I've already done and I want help with the implementation strategy.

- [x] use generics for I2C
- [x] use generics for SmBus
- [x] implement a `NoSMBus` type
- [x] make it compile for all the boards
- [x] observe de binary sizes for imix, microbit and risc-v board

### Documentation Updated

- [ ] No updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Comment by alexandruradovici** [comment]
It works, we have to change `&dyn I2CDevice` to use a trait bound.

**Comment by JADarius** [comment]
I'm sorry, but I don't understand what I have to do.

**Comment by alexandruradovici** [comment]
You must modify all the usages of `&dyn I2CDevice` to to use a trait bound `I: I2CDevice`.

**Comment by JADarius** [comment]
Where should I look for `&dyn I2CDevice`? I can't find the files that have that `&dyn I2CDevice`.

**Comment by alexandruradovici** [comment]
Just some examples:

https://github.com/UPB-CS-OpenSourceUpstream/tock/blob/6f91e8df807976582513ff62e571622ef4df66b6/capsules/src/lsm303agr.rs#L145-L146

https://github.com/search?q=repo%3AUPB-CS-OpenSourceUpstream%2Ftock%20%26%27a%20dyn%20i2c%3A%3AI2CDevice&type=code

**Comment by JADarius** [comment]
I changed all the `&dyn I2CDevice` that I've found, but I get a compilation error in all of the components affected by the change: `expected type parameter 'J', found struct 'capsules::virtual_i2c::I2CDevice` and I don't understand where I made the mistake.

**Comment by alexandruradovici** [comment]
The problem is the following: you have two types:
- a trait called `I2CDevice` defined in https://github.com/UPB-CS-OpenSourceUpstream/tock/blob/1be6f52a6da75276dff24b61b30ae70a56c6f198/kernel/src/hil/i2c.rs#L217
- a structure that implements a virtual `I2CDevice` defined in https://github.com/UPB-CS-OpenSourceUpstream/tock/blob/1be6f52a6da75276dff24b61b30ae70a56c6f198/capsules/src/virtual_i2c.rs#L211

You need to be careful which one is used. The `adps9960` driver uses the `I2CDevice` structure, which in terms uses the `I2CDevice` trait. You do not need the extra generic here, because you know the actual data type that you will use (the `I2CDevice` structure). The only template that you need is the `I2CMaster`, which you have correctly used.

https://github.com/UPB-CS-OpenSourceUpstream/tock/blob/1be6f52a6da75276dff24b61b30ae70a56c6f198/capsules/src/virtual_i2c.rs#L254

When  it comes it the actual sensor, your modifications are correct, it receives any structure that implements the `I2CDevice` trait. The component will use a `I2CDevice<'static, I>` structure, which does implement the `I2CDevice` trait.

File: boards/components/src/apds9960.rs [hunk 1]
```diff
@@ -2,48 +2,54 @@
...
 }
 
-pub struct Apds9960Component {
-    i2c_mux: &'static MuxI2C<'static>,
+pub struct Apds9960Component<I: 'static + i2c::I2CMaster, J: 'static + i2c::I2CDevice> {
+    i2c_mux: &'static MuxI2C<'static, I>,
     i2c_address: u8,
     interrupt_pin: &'static dyn gpio::InterruptPin<'static>,
+    i2c_device: PhantomData<J>,
 }
```
Comment (alexandruradovici) [inline boards/components/src/apds9960.rs]:
```suggestion
pub struct Apds9960Component<I: 'static + i2c::I2CMaster> {
    i2c_mux: &'static MuxI2C<'static, I>,
    i2c_address: u8,
    interrupt_pin: &'static dyn gpio::InterruptPin<'static>,
}
```

File: boards/components/src/apds9960.rs [hunk 2]
```diff
@@ -2,48 +2,54 @@
...
 
-impl Apds9960Component {
+impl<I: 'static + i2c::I2CMaster, J: 'static + i2c::I2CDevice> Apds9960Component<I, J> {
```
Comment (alexandruradovici) [inline boards/components/src/apds9960.rs]:
```suggestion
impl<I: 'static + i2c::I2CMaster> Apds9960Component<I> {
```

File: boards/components/src/apds9960.rs [hunk 3]
```diff
@@ -2,48 +2,54 @@
...
-        i2c_mux: &'static MuxI2C<'static>,
+        i2c_mux: &'static MuxI2C<'static, I>,
         i2c_address: u8,
         interrupt_pin: &'static dyn gpio::InterruptPin<'static>,
     ) -> Self {
         Apds9960Component {
             i2c_mux,
             i2c_address,
             interrupt_pin,
+            i2c_device: PhantomData,
```
Comment (alexandruradovici) [inline boards/components/src/apds9960.rs]:
```suggestion
```

File: boards/components/src/apds9960.rs [hunk 4]
```diff
@@ -2,48 +2,54 @@
...
         }
     }
 }
 
-impl Component for Apds9960Component {
+impl<I: 'static + i2c::I2CMaster, J: 'static + i2c::I2CDevice> Component
+    for Apds9960Component<I, J>
+{
```
Comment (alexandruradovici) [inline boards/components/src/apds9960.rs]:
```suggestion
impl<I: 'static + i2c::I2CMaster> Component
    for Apds9960Component<I>
{
```

File: boards/components/src/apds9960.rs [hunk 5]
```diff
@@ -2,48 +2,54 @@
...
     type StaticInput = (
-        &'static mut MaybeUninit<I2CDevice<'static>>,
-        &'static mut MaybeUninit<APDS9960<'static>>,
+        &'static mut MaybeUninit<I2CDevice<'static, I>>,
+        &'static mut MaybeUninit<APDS9960<'static, J>>,
```
Comment (alexandruradovici) [inline boards/components/src/apds9960.rs]:
```suggestion
        &'static mut MaybeUninit<APDS9960<'static, I2CDevice<'static, I>>>,
```

This is the mistake, the component uses the `I2CDevice` structure.

File: boards/components/src/apds9960.rs [hunk 6]
```diff
@@ -2,48 +2,54 @@
...
         &'static mut MaybeUninit<[u8; capsules::apds9960::BUF_LEN]>,
     );
-    type Output = &'static APDS9960<'static>;
+    type Output = &'static APDS9960<'static, J>;
```
Comment (alexandruradovici) [inline boards/components/src/apds9960.rs]:
```suggestion
    type Output = &'static APDS9960<'static, I2CDevice<'static, I>>;
```

File: boards/components/src/apds9960.rs [hunk 7]
```diff
@@ -2,48 +2,54 @@
 
 use capsules::apds9960::APDS9960;
 use capsules::virtual_i2c::{I2CDevice, MuxI2C};
+use core::marker::PhantomData;
```
Comment (alexandruradovici) [inline boards/components/src/apds9960.rs]:
```suggestion
```

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

**Comment by JADarius** [comment]
I have a question: how do I implement the static components macros? Right now they give me errors when I try to compile for the microbit.

**Comment by alexandruradovici** [comment]
You have to add a parameter for the I2C type to every component that uses the I2C. Take a look at components that use SPI. The difference is that SPI has two generic types.

**Comment by JADarius** [comment]
The code compiles for the microbit now, but I'm not sure that I have made the correct modifications, especially in the microbit's `main.rs`. How do I test if the I2C works?

**Comment by alexandruradovici** [comment]
The easiest way to test is to run the sensors application. If it displays the motion sensor, it means it should work fine.

**Comment by JADarius** [comment]
I've run the sensors application and it gave me the following output:
`Temperature:                 25 deg C`
`Acceleration: X: -19 Y: -750 Z: 683`
`Magnetometer: X: 3742 Y: -1516 Z: 389`
`Sound Pressure:              67`
I've also run the sensors application on an unmodified version of Tock and it looks the same.

**Comment by JADarius** [comment]
I have also checked the sizes of `microbit_v2` from both of the versions. The unmodified one has 4225184 bytes and the one that uses generics has 4203324 bytes.

**Comment by alexandruradovici** [comment]
This seems to work. We still need to fix the SMBus in https://github.com/UPB-CS-OpenSourceUpstream/tock/blob/1be6f52a6da75276dff24b61b30ae70a56c6f198/capsules/src/virtual_i2c.rs#L17.

This is a different type of I2C. The problem is that it is in an `Option`, so it is not always used. This means that for boards that do not have SMBus, like the microbit v2, there is no way to supply an actual type.

My suggestion is to define a `NoSMBus` type like:

```rust
struct NoSMBus;

impl SMBusMaster for NoSMBus {
  // all functions return Err(ErrorCode::NOSUPPORT)
}
```

in https://github.com/UPB-CS-OpenSourceUpstream/tock/blob/master/kernel/src/hil/i2c.rs.

Use this type as the default type for the smbus like:

```rust
pub struct MuxI2C<'a, I: i2c::I2CMaster, S: i2c::SMBusMaster = NoSMBus> {
    i2c: &'a I,
    smbus: Option<&'a S>,
   // ...
}
```

File: capsules/src/virtual_i2c.rs [hunk 8]
```diff
@@ -305,18 +305,18 @@ impl i2c::I2CDevice for I2CDevice<'_> {
     }
 }
 
-pub struct SMBusDevice<'a> {
-    mux: &'a MuxI2C<'a>,
+pub struct SMBusDevice<'a, I: i2c::I2CMaster> {
```
Comment (alexandruradovici) [inline capsules/src/virtual_i2c.rs]:
```suggestion
pub struct SMBusDevice<'a, I: i2c::I2CMaster, S: i2c::SMBusMaster> {
```

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

**Comment by alexandruradovici** [comment]
# Tasks
- [ ] use generics for I2C
- [ ] use generics for SmBus
- [ ] implement a `NoSMBus` type
- [ ] make it compile for all the boards
- [ ] observe de binary sizes for imix, microbit and risc-v board

File: capsules/src/virtual_i2c.rs [hunk 9]
```diff
@@ -193,7 +193,7 @@ impl<'a, I: i2c::I2CMaster> MuxI2C<'a, I> {
     }
 }
 
-impl<'a, I: i2c::I2CMaster> DynamicDeferredCallClient for MuxI2C<'a, I> {
+impl<'a, I: i2c::I2CMaster, S: i2c::SMBusMaster> DynamicDeferredCallClient for MuxI2C<'a, I, S> {
```
Comment (alexandruradovici) [inline capsules/src/virtual_i2c.rs]:
```suggestion
impl<'a, I: i2c::I2CMaster, S: i2c::SMBusMaster = NoSMBus> DynamicDeferredCallClient for MuxI2C<'a, I, S> {
```

Comment (JADarius) [reply capsules/src/virtual_i2c.rs]: This gives me the following error: `error: defaults for type parameters are only allowed in 'struct', 'enum', 'type', or 'trait' definitions`

File: capsules/src/virtual_i2c.rs [hunk 10]
```diff
@@ -305,18 +305,18 @@ impl i2c::I2CDevice for I2CDevice<'_> {
...
     buffer: TakeCell<'static, [u8]>,
     operation: Cell<Op>,
-    next: ListLink<'a, SMBusDevice<'a>>,
+    next: ListLink<'a, SMBusDevice<'a, I>>,
     client: OptionalCell<&'a dyn I2CClient>,
 }
 
-impl<'a> SMBusDevice<'a> {
-    pub fn new(mux: &'a MuxI2C<'a>, addr: u8) -> SMBusDevice<'a> {
+impl<'a, I: i2c::I2CMaster> SMBusDevice<'a, I> {
```
Comment (alexandruradovici) [inline capsules/src/virtual_i2c.rs]:
```suggestion
impl<'a, I: i2c::I2CMaster, S: i2c::SMBusMaster> SMBusDevice<'a, I, S> {
```

File: capsules/src/virtual_i2c.rs [hunk 11]
```diff
@@ -305,18 +305,18 @@ impl i2c::I2CDevice for I2CDevice<'_> {
...
-pub struct SMBusDevice<'a> {
-    mux: &'a MuxI2C<'a>,
+pub struct SMBusDevice<'a, I: i2c::I2CMaster> {
+    mux: &'a MuxI2C<'a, I>,
     addr: u8,
     enabled: Cell<bool>,
     buffer: TakeCell<'static, [u8]>,
     operation: Cell<Op>,
-    next: ListLink<'a, SMBusDevice<'a>>,
+    next: ListLink<'a, SMBusDevice<'a, I>>,
```
Comment (alexandruradovici) [inline capsules/src/virtual_i2c.rs]:
```suggestion
    next: ListLink<'a, SMBusDevice<'a, I, S>>,
```

File: capsules/src/virtual_i2c.rs [hunk 12]
```diff
@@ -9,22 +9,22 @@ use kernel::collections::list::{List, ListLink, ListNode};
...
-pub struct MuxI2C<'a> {
-    i2c: &'a dyn i2c::I2CMaster,
-    smbus: Option<&'a dyn i2c::SMBusMaster>,
-    i2c_devices: List<'a, I2CDevice<'a>>,
-    smbus_devices: List<'a, SMBusDevice<'a>>,
+pub struct MuxI2C<'a, I: i2c::I2CMaster, S: i2c::SMBusMaster = NoSMBus> {
+    i2c: &'a I,
+    smbus: Option<&'a S>,
+    i2c_devices: List<'a, I2CDevice<'a, I>>,
+    smbus_devices: List<'a, SMBusDevice<'a, I>>,
```
Comment (alexandruradovici) [inline capsules/src/virtual_i2c.rs]:
```suggestion
    smbus_devices: List<'a, SMBusDevice<'a, S>>,
```

File: capsules/src/virtual_i2c.rs [hunk 13]
```diff
@@ -9,22 +9,22 @@ use kernel::collections::list::{List, ListLink, ListNode};
...
     enabled: Cell<usize>,
-    i2c_inflight: OptionalCell<&'a I2CDevice<'a>>,
-    smbus_inflight: OptionalCell<&'a SMBusDevice<'a>>,
+    i2c_inflight: OptionalCell<&'a I2CDevice<'a, I>>,
+    smbus_inflight: OptionalCell<&'a SMBusDevice<'a, I>>,
```
Comment (alexandruradovici) [inline capsules/src/virtual_i2c.rs]:
```suggestion
    smbus_inflight: OptionalCell<&'a SMBusDevice<'a, S>>,
```

File: capsules/src/virtual_i2c.rs (hunk 9 above)
Comment (alexandruradovici) [reply capsules/src/virtual_i2c.rs]: Yes, that is correct. Do not modify it.

File: capsules/src/virtual_i2c.rs [hunk 14]
```diff
@@ -305,18 +305,18 @@ impl i2c::I2CDevice for I2CDevice<'_> {
     }
 }
 
-pub struct SMBusDevice<'a> {
-    mux: &'a MuxI2C<'a>,
+pub struct SMBusDevice<'a, I: i2c::I2CMaster> {
+    mux: &'a MuxI2C<'a, I>,
```
Comment (alexandruradovici) [inline capsules/src/virtual_i2c.rs]:
```suggestion
    mux: &'a MuxI2C<'a, I, S>,
```

File: capsules/src/virtual_i2c.rs [hunk 15]
```diff
@@ -305,18 +305,18 @@ impl i2c::I2CDevice for I2CDevice<'_> {
...
     operation: Cell<Op>,
-    next: ListLink<'a, SMBusDevice<'a>>,
+    next: ListLink<'a, SMBusDevice<'a, I>>,
     client: OptionalCell<&'a dyn I2CClient>,
 }
 
-impl<'a> SMBusDevice<'a> {
-    pub fn new(mux: &'a MuxI2C<'a>, addr: u8) -> SMBusDevice<'a> {
+impl<'a, I: i2c::I2CMaster> SMBusDevice<'a, I> {
+    pub fn new(mux: &'a MuxI2C<'a, I>, addr: u8) -> SMBusDevice<'a, I> {
```
Comment (alexandruradovici) [inline capsules/src/virtual_i2c.rs]:
```suggestion
    pub fn new(mux: &'a MuxI2C<'a, I, S>, addr: u8) -> SMBusDevice<'a, I, S> {
```

File: capsules/src/virtual_i2c.rs [hunk 16]
```diff
@@ -338,21 +338,21 @@ impl<'a> SMBusDevice<'a> {
     }
 }
 
-impl<'a> I2CClient for SMBusDevice<'a> {
+impl<'a, I: i2c::I2CMaster> I2CClient for SMBusDevice<'a, I> {
```
Comment (alexandruradovici) [inline capsules/src/virtual_i2c.rs]:
```suggestion
impl<'a, I: i2c::I2CMaster, S: i2c::SMBusMaster> I2CClient for SMBusDevice<'a, I, S> {
```

File: capsules/src/virtual_i2c.rs [hunk 17]
```diff
@@ -208,18 +208,18 @@ enum Op {
     CommandComplete(Result<(), Error>),
 }
 
-pub struct I2CDevice<'a> {
-    mux: &'a MuxI2C<'a>,
+pub struct I2CDevice<'a, I: i2c::I2CMaster> {
+    mux: &'a MuxI2C<'a, I>,
```
Comment (alexandruradovici) [inline capsules/src/virtual_i2c.rs]:
```suggestion
pub struct I2CDevice<'a, I: i2c::I2CMaster, S: i2c::SMBusMaster = NoSMBus> {
    mux: &'a MuxI2C<'a, I, S>,
```

File: capsules/src/virtual_i2c.rs [hunk 18]
```diff
@@ -208,18 +208,18 @@ enum Op {
...
     addr: u8,
     enabled: Cell<bool>,
     buffer: TakeCell<'static, [u8]>,
     operation: Cell<Op>,
-    next: ListLink<'a, I2CDevice<'a>>,
+    next: ListLink<'a, I2CDevice<'a, I>>,
```
Comment (alexandruradovici) [inline capsules/src/virtual_i2c.rs]:
```suggestion
    next: ListLink<'a, I2CDevice<'a, I, S>>,
```

File: capsules/src/virtual_i2c.rs [hunk 19]
```diff
@@ -208,18 +208,18 @@ enum Op {
...
     client: OptionalCell<&'a dyn I2CClient>,
 }
 
-impl<'a> I2CDevice<'a> {
-    pub fn new(mux: &'a MuxI2C<'a>, addr: u8) -> I2CDevice<'a> {
+impl<'a, I: i2c::I2CMaster> I2CDevice<'a, I> {
```
Comment (alexandruradovici) [inline capsules/src/virtual_i2c.rs]:
```suggestion
impl<'a, I: i2c::I2CMaster, S: i2c::SMBusMaster> I2CDevice<'a, I, S> {
```

File: capsules/src/virtual_i2c.rs [hunk 20]
```diff
@@ -208,18 +208,18 @@ enum Op {
...
+    pub fn new(mux: &'a MuxI2C<'a, I>, addr: u8) -> I2CDevice<'a, I> {
```
Comment (alexandruradovici) [inline capsules/src/virtual_i2c.rs]:
```suggestion
    pub fn new(mux: &'a MuxI2C<'a, I, S>, addr: u8) -> I2CDevice<'a, I, S> {
```

File: capsules/src/virtual_i2c.rs [hunk 21]
```diff
@@ -237,21 +237,21 @@ impl<'a> I2CDevice<'a> {
     }
 }
 
-impl I2CClient for I2CDevice<'_> {
+impl<I: i2c::I2CMaster> I2CClient for I2CDevice<'_, I> {
```
Comment (alexandruradovici) [inline capsules/src/virtual_i2c.rs]:
```suggestion
impl<I: i2c::I2CMaster, S: i2c::SMBusMaster> I2CClient for I2CDevice<'_, I, S> {
```

File: capsules/src/virtual_i2c.rs [hunk 22]
```diff
@@ -338,21 +338,21 @@ impl<'a> SMBusDevice<'a> {
...
     fn command_complete(&self, buffer: &'static mut [u8], status: Result<(), Error>) {
         self.client.map(move |client| {
             client.command_complete(buffer, status);
         });
     }
 }
 
-impl<'a> ListNode<'a, SMBusDevice<'a>> for SMBusDevice<'a> {
-    fn next(&'a self) -> &'a ListLink<'a, SMBusDevice<'a>> {
+impl<'a, I: i2c::I2CMaster> ListNode<'a, SMBusDevice<'a, I>> for SMBusDevice<'a, I> {
```
Comment (alexandruradovici) [inline capsules/src/virtual_i2c.rs]:
```suggestion
impl<'a, I: i2c::I2CMaster, S: i2c::SMBusMaster> ListNode<'a, SMBusDevice<'a, I, S>> for SMBusDevice<'a, I, S> {
```

**Review by alexandruradovici — CHANGES_REQUESTED** [review]
I modified the templates a little bit, you should know get an idea how to fix the others.

**Comment by JADarius** [comment]
I made the modifications and I changed the main.rs to some boards to that they work but I have a problem: for the `stm32f3discovery` board, I have to add a `I2CDevice` component for the `Lsm303dlhcI2C` on line 68.

**Comment by JADarius** [comment]
I modified the code and I think that it compiles for all the boards (`make prepush` says that all checks passed). I had one problem: the nRF52840 board needed a Master Slave interface and TWI didn't explicitly implement the `hil::i2c::I2CMasterSlave` trait so I did a hack where I implement it in the implementation of I2C for nRF52 and I don't know if it was a good idea.

**Comment by alexandruradovici** [comment]
Looks good. Can you check the code size (printed by make) for a few boards?

**Comment by JADarius** [comment]
Old imix:
`   text    data     bss     dec     hex ` 
` 175092       0   28568  203660   31b8c `
New imix:
`   text    data     bss     dec     hex `
` 174836       0   28556  203392   31a80 `

Old microbit:
`   text    data     bss     dec     hex `
` 106500       0   15788  122288   1ddb0 `
New microbit:
`   text    data     bss     dec     hex `
` 106500       0   15772  122272   1dda0 `

Old stm32f3discovery:
`   text    data     bss     dec     hex ` 
` 100356       0   16632  116988   1c8fc `
New stm32f3discovery:
`   text    data     bss     dec     hex `
` 100356       0   16616  116972   1c8ec `

**Comment by alexandruradovici** [comment]
Can you check the code sizes for a risc-v board?

File: kernel/src/hil/i2c.rs [hunk 23]
```diff
@@ -177,7 +177,7 @@ pub trait I2CSlave {
 /// Master and Slave modes.
 pub trait I2CMasterSlave: I2CMaster + I2CSlave {}
 // Provide blanket implementations for trait group
-impl<T: I2CMaster + I2CSlave> I2CMasterSlave for T {}
+// impl<T: I2CMaster + I2CSlave> I2CMasterSlave for T {}
```
Comment (alexandruradovici) [inline kernel/src/hil/i2c.rs:180]: Why is this commented now?

File: capsules/src/virtual_i2c.rs [hunk 24]
```diff
@@ -9,22 +9,22 @@ use kernel::collections::list::{List, ListLink, ListNode};
...
-use kernel::hil::i2c::{self, Error, I2CClient, I2CHwMasterClient};
+use kernel::hil::i2c::{self, Error, I2CClient, I2CHwMasterClient, NoSMBus};
 use kernel::utilities::cells::{OptionalCell, TakeCell};
 
-pub struct MuxI2C<'a> {
-    i2c: &'a dyn i2c::I2CMaster,
-    smbus: Option<&'a dyn i2c::SMBusMaster>,
-    i2c_devices: List<'a, I2CDevice<'a>>,
-    smbus_devices: List<'a, SMBusDevice<'a>>,
+pub struct MuxI2C<'a, I: i2c::I2CMaster, S: i2c::SMBusMaster = NoSMBus> {
```
Comment (alexandruradovici) [inline capsules/src/virtual_i2c.rs]: Add a comment explaining why this structure is necessary.

File: boards/components/src/i2c.rs [hunk 25]
```diff
@@ -21,33 +21,33 @@ use capsules::virtual_i2c::{I2CDevice, MuxI2C};
...
 use kernel::dynamic_deferred_call::DynamicDeferredCall;
-use kernel::hil::i2c;
+use kernel::hil::i2c::{self, NoSMBus};
 
 // Setup static space for the objects.
 #[macro_export]
 macro_rules! i2c_mux_component_static {
-    () => {{
-        kernel::static_buf!(capsules::virtual_i2c::MuxI2C<'static>)
+    ($I:ty $(,)?) => {{
```
Comment (alexandruradovici) [inline boards/components/src/i2c.rs:28]: Add another macro branch that receives a SMBus type.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

**Comment by JADarius** [comment]
I chose a `hifive` board to test the sizes.
Unmodified version:
`   text    data     bss     dec     hex `
`  77388      12    7320   84720   14af0 `
Modified version:
`   text    data     bss     dec     hex `
`  75276      12    6932   82220   1412c `

File: kernel/src/hil/i2c.rs (hunk 23 above)
Comment (JADarius) [reply kernel/src/hil/i2c.rs:180]: I commented this because it conflicts the implementation from `chips/nrf52/src/i2c.rs` at line 364.

**Comment by alexandruradovici** [comment]
I suggest rebasing this, fixing the errors and sending it to upstream. Please write a PR message explaining all the modifications.

**Comment by alexandruradovici** [comment]
@JADarius any progress here?

**Comment by alexandruradovici** [comment]
@JADarius ping?

**Comment by JADarius** [comment]
I rebased it and solved the problems with my code but `make prepush` gives me an error about a function I didn't touch. The error is in the file `kernel/src/process_standard.rs` on line 246.

**Review by alexandruradovici — APPROVED** [review]
Send this pull request to upstream with a message similar to:

### Pull Request Overview

This pull request modifies the I2C HIL and capsules to use generics.

##### Code sizes

###### Old imix:
`   text    data     bss     dec     hex ` 
` 175092       0   28568  203660   31b8c `
###### New imix:
`   text    data     bss     dec     hex `
` 174836       0   28556  203392   31a80 `

###### Old microbit:
`   text    data     bss     dec     hex `
` 106500       0   15788  122288   1ddb0 `
###### New microbit:
`   text    data     bss     dec     hex `
` 106500       0   15772  122272   1dda0 `

###### Old stm32f3discovery:
`   text    data     bss     dec     hex ` 
` 100356       0   16632  116988   1c8fc `
###### New stm32f3discovery:
`   text    data     bss     dec     hex `
` 100356       0   16616  116972   1c8ec `


###### HiFive
Unmodified version:
`   text    data     bss     dec     hex `
`  77388      12    7320   84720   14af0 `
Modified version:
`   text    data     bss     dec     hex `
`  75276      12    6932   82220   1412c `

##### Pros
- this allows the compiler to perform better optimization and function inlining;
- smaller code size.
##### Cons
- writing I2C drivers is more verbose as data types are longer;
- `SMBus` was optional, this uses an "empty" implementation.

### TODO or Help Wanted

Feedback would be highly appreciated.

### Documentation Updated

- [ ] No updates are required.

### Formatting

- [x] Ran `make prepush`.

**Comment by JADarius** [comment]
I have made a PR: https://github.com/tock/tock/pull/3431

**Pull request merged**